app/models/spree/gateway/amazon.rb

```diff
@@ -183,6 +183,16 @@ def void(response_code, gateway_options)
       return ActiveMerchant::Billing::Response.new(true, "Success", response)
     end
 
+    def cancel(response_code)
+      payment = Spree::Payment.find_by(response_code: response_code)
+      order = payment.order
+      load_amazon_mws(payment.source.order_reference)
+
+      response = @mws.refund(response_code, order.number, payment.credit_allowed, payment.currency)
```

Comment:
> all we are given is response code, how do we know if we have to cancel or refund?

We can do the same thing that `def void` is doing, just above this `def cancel` right?

> The cancel! method is only called on payments that have been captured

Great catch about [Order#after_cancel](https://github.com/solidusio/solidus/blob/ebda5c8/core/app/models/spree/order.rb#L762-L769).  I think that might be a bug in Solidus.  I've opened a PR to find out:

Either way, `Payment#cancel` is an interface that is callable outside of `Order#after_cancel`.  Shouldn't calling `cancel` on an authorized Amazon payment cause it to be voided?  Other Solidus code I can find does that.  E.g. [the Braintree example](https://github.com/solidusio/solidus_braintree/blob/81176dc/app/models/solidus/gateway/braintree_gateway.rb#L169-L177) I mentioned and also [StoreCredit in Solidus](https://github.com/solidusio/solidus/blob/v2.0.0/core/app/models/spree/payment_method/store_credit.rb#L81-L85).